internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +100,210 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
+	var clientID, clientSecret string
+	secret, applicationErr := c.secretInformer.Lister().Secrets(upstream.Namespace).Get(upstream.Spec.Client.SecretName)
+	if applicationErr != nil {
+		applicationErrors = append(applicationErrors, applicationErr)
+	}
+	if applicationErr != nil ||
+		secret == nil ||
+		secret.Type != gitHubClientSecretType ||
+		len(secret.Data[clientIDDataKey]) < 1 ||
+		len(secret.Data[clientSecretDataKey]) < 1 {
+		conditions = append(conditions, &metav1.Condition{
+			Type:   ClientCredentialsObtained,
+			Status: metav1.ConditionFalse,
+			Reason: upstreamwatchers.ReasonNotFound,
+			Message: fmt.Sprintf("secret with name %q must be found in namespace %q with type %q and keys %q and %q",
+				upstream.Spec.Client.SecretName,
+				upstream.Namespace,
+				gitHubClientSecretType,
+				clientIDDataKey,
+				clientSecretDataKey),
+		})
+	} else if applicationErr == nil {
+		clientID = string(secret.Data[clientIDDataKey])
+		clientSecret = string(secret.Data[clientSecretDataKey])
+		conditions = append(conditions, &metav1.Condition{
+			Type:    ClientCredentialsObtained,
+			Status:  metav1.ConditionTrue,
+			Reason:  upstreamwatchers.ReasonSuccess,
+			Message: "client credentials have been obtained, although they may not yet have been validated",
+		})
+	}
+
+	// Should not happen due to CRD defaulting
+	host := ""
+	if upstream.Spec.GitHubAPI.Host == nil || len(*upstream.Spec.GitHubAPI.Host) < 1 {
+		conditions = append(conditions, &metav1.Condition{
+			Type:    HostValid,
+			Status:  metav1.ConditionFalse,
+			Reason:  "InvalidHost",
+			Message: `spec.githubAPI.host ("") is not valid`,
+		})
+		conditions = append(conditions, &metav1.Condition{
+			Type:    GitHubConnectionValid,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "UnableToValidate",
+			Message: `unable to determine if spec.githubAPI.host ("") is reachable or if TLS verification succeeds`,
+		})
+		conditions = append(conditions, &metav1.Condition{
+			Type:    TLSConfigurationValid,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "UnableToValidate",
+			Message: `tls configuration cannot be validated for spec.githubAPI.host ("")`,
+		})
+	} else {
+		host = *upstream.Spec.GitHubAPI.Host
+		conditions = append(conditions, &metav1.Condition{
+			Type:    HostValid,
+			Status:  metav1.ConditionTrue,
+			Reason:  upstreamwatchers.ReasonSuccess,
+			Message: fmt.Sprintf("spec.githubAPI.host (%q) is valid", host),
+		})
+	}
+
+	// Should not happen due to CRD defaulting
+	if upstream.Spec.Claims.Groups == nil || upstream.Spec.Claims.Username == nil {
+		return nil, fmt.Errorf("spec.claims.groups and spec.claims.username are required")
+	}
+
+	// Should not happen due to CRD enum validation
+	switch *upstream.Spec.Claims.Username {
+	case v1alpha1.GitHubUsernameLoginAndID:
+	case v1alpha1.GitHubUsernameLogin:
+	case v1alpha1.GitHubUsernameID:
+		break
+	default:
+		return nil, fmt.Errorf("invalid spec.claims.username")
```

Comment:
If it really can't happen, do we just want to `log` the error?
If an error here does happen, then it is a `user configuration` error, so we don't really want to return a `Sync()` error and trigger the loop again.  A user has to step in and fix the config.